pkg/mutation/match.go

```diff
@@ -0,0 +1,136 @@
+package mutation
+
+import (
+	"fmt"
+
+	mutationsv1 "github.com/open-policy-agent/gatekeeper/apis/mutations/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// Matches verifies if the given object belonging to the given namespace
+// matches the current mutator.
+func Matches(match mutationsv1.Match, obj *unstructured.Unstructured, ns *corev1.Namespace) (bool, error) {
+	meta, err := meta.Accessor(obj)
+	if err != nil {
+		return false, fmt.Errorf("Accessor failed for %s", obj.GetObjectKind().GroupVersionKind().Kind)
+	}
+
+	foundMatch := false
+
+	for _, kk := range match.Kinds {
+		kindMatches := false
+		groupMatches := false
+
+		for _, k := range kk.Kinds {
+			if k == "*" || k == obj.GroupVersionKind().Kind {
+				kindMatches = true
+				break
+			}
+		}
+		for _, g := range kk.APIGroups {
+			if g == "*" || g == obj.GroupVersionKind().Group {
+				groupMatches = true
+				break
+			}
+		}
+		if kindMatches && groupMatches {
+			foundMatch = true
+		}
+	}
+	if len(match.Kinds) == 0 {
+		foundMatch = true
+	}
+
+	if !foundMatch {
+		return false, nil
+	}
+
+	if match.Scope == apiextensionsv1beta1.ClusterScoped &&
+		meta.GetNamespace() != "" {
+		return false, nil
+	}
+
+	if match.Scope == apiextensionsv1beta1.NamespaceScoped &&
+		meta.GetNamespace() == "" {
+		return false, nil
+	}
+
+	found := false
+	for _, n := range match.Namespaces {
+		if meta.GetNamespace() == n {
+			found = true
+			break
+		}
+	}
+	if !found && len(match.Namespaces) > 0 {
+		return false, nil
+	}
+
+	for _, n := range match.ExcludedNamespaces {
+		if meta.GetNamespace() == n {
+			return false, nil
+		}
+	}
+	if match.LabelSelector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(match.LabelSelector)
+		if err != nil {
+			return false, err
+		}
+		if !selector.Matches(labels.Set(meta.GetLabels())) {
+			return false, nil
+		}
+	}
+
+	if match.NamespaceSelector != nil {
```

Comment:
For MutatingWebhookConfigurations:

* Cluster scoped resources always match an NS selector
* If the object itself is an NS, the namespace selector is evaluated against the object's labels
* Unclear whether the old object + new object matching logic also applies here

Reply:
The namespace remains the same, so there won't be difference in new vs old.

Reply:
Implemented the first two points and added tests.

Reply:
The namespace would remain the same unless the object itself is a namespace.